src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs

```diff
@@ -49,6 +49,8 @@ internal class PrioritySendDataCollection
         // fragmentor used to serialize & fragment objects added to this collection.
         private Fragmentor _fragmentor;
         private object[] _syncObjects;
+        // synchronize access to _syncObjects and _dataToBeSent as a set.
+        private object _syncObject = new Object();
```

Comment:
This field is different from array "_syncObjects" only by last letter - it is confusing.
It would be good to either add comment for "_syncObjects" array about what it is; or rename "_syncObject"; or both.

Reply:
@anmenaga: Actually, after reviewing this logic again, _syncObject is insufficient since it only guards Clear(). Additionally, _syncObjects also appears to be fragile; it guards against individual usage of a given element but assumes the elements are valid; which is the actual problem.

My thinking is _syncObjects should be removed and _syncObject used to serialize access to the set instead of individual members.

Thoughts?

Reply:
@dantraMSFT I'm all for simplification, but there should be a reason why they decided to go more complex 'individual-sync-object' route in the first place... Maybe something will deadlock with a single syncObject, and parallel "Add"s won't be able to execute for different DataPriorityTypes (not sure how critical this really is).

Reply:
> Actually, after reviewing this logic again, _syncObject is insufficient since it only guards Clear().

`_dataSyncObjects` and `_dataToBeSent` are used in multiple methods, but it appears that there is an assumption that those methods won't be called until the `Fragmentor` property is set (where both `_dataSyncObjects` and `_dataToBeSent` would be initialized). With this assumption, I think it's fine to directly use `_dataSyncObjects` and `_dataToBeSent` in methods other than `Clear()`. But we probably should add a comment about that assumption.

For `Clear()`, it's possible to be called before `Fragmentor` being set (in error path), so it's necessary to check `_dataSyncObjects` for null beforehand.